clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1541,14 +1574,31 @@ public void commitAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, Of
         acquireAndEnsureOpen();
         try {
             maybeThrowInvalidGroupIdException();
+            maybeThrowIfCommitOffsetsNotOwned(offsets);
             log.debug("Committing offsets: {}", offsets);
             offsets.forEach(this::updateLastSeenEpochIfNewer);
             coordinator.commitOffsetsAsync(new HashMap<>(offsets), callback);
+        } catch (CommitFailedException e) {
+            log.error("Failed to commit offsets asynchronously because they do not belong to dynamically assigned partitions");
+            callback.onComplete(offsets, e);
         } finally {
             release();
         }
     }
 
+    // this is an optimization on the client side: if we are doomed to be rejected by the group coordinator
+    // due to invalid generation, we can just fail fast and inform the callers earlier
+    private void maybeThrowIfCommitOffsetsNotOwned(final Map<TopicPartition, OffsetAndMetadata> offsets) {
```

Comment:
nit: The callers of this iterate the committed offsets twice. Once in order to check partition ownership and then a second time to invoke `updateLastSeenEpochIfNewer`. I wonder if we could consolidate these two loops.

Reply:
Yes, if we are not gonna remove this (see other comment above) I can do this.